### ISSUE TITLE
Some of us are noobs

### ISSUE THREAD

**Issue by authenticastrology** [lilyinstarlight/nixos-cosmic#404]
Some of us...well okay... at least myself are not fully adept with flakes and it is like rocket science to us. But, we want to install Cosmic anyway. Would love a step-by-step guide. Know you may not have the time. But if anyone else does, it would be excellent. Or a video. But knowing whether to add this to the default flake, and how...Or whether to create a new one...would be excellent.

Sorry if this is the wrong spot for this. But nowhere else to comment.

**Comment by colemickens** [comment]
@authenticastrology The README generally demonstrates what you'd need to add to your flake to add `nixos-cosmic` as an input to your flake, as well as how to utilize it into your system configuration.

For example:

>to the default flake, and how...Or whether to create a new one...

is a bit confusing, or leads me to wonder if you have some confusion?

For example, I have a single flake that encompasses:
- my laptop nixos config (which utilizes the `nixos-cosmic` nixos module)
- my server nixos config
- devshells for various "things"
- various packages that I have pinned to tip versions

That singular flake is where I added `nixos-cosmic` as an input, and then in one of my system configurations, I pull in the nixos module, from that flake input.

**Comment by Purian23** [comment]
@authenticastrology, were you able to get sorted to get up and running?

**Comment by authenticastrology** [comment]
No, not yet. But, thank you.

**Comment by FelixSchausberger** [comment]
Can you tell where you're stuck? Or provide your dotfiles?

What always helped me a lot in the beginning is to see how other people have implemented it, so [here](https://github.com/FelixSchausberger/nixos) are my dotfiles if you want to take a look at them.

**Comment by foldfree** [comment]
i just got it working yesterday. I recommend you spend some time reading the [the wiki on flakes](https://nixos.wiki/wiki/flakes) + get help from claude.ai and phind.com or whatever LLM you like for troubleshooting.

one thing missing in the readme (I think?) is that you should do ` sudo nixos-rebuild test --flake .` and not just `nixos rebuild test` before adding the `cosmic enable` in your configuration.nix.
Then rebuid with `sudo nixos-rebuild switch --flake .`

Also make sure you did not add any `nixpkgs` input in your flakes inputs beside the cosmic one to benefit from cachix.
I’m quite clueless about nixos this took me a couple of hours. My config had some cleanup to do. If you start from a vanilla install, it should be easier.

**Comment by authenticastrology** [comment]
I am almost embarrassed to ask. I am still just barely grasping flakes. I
would be really grateful if someone has a pre-made template just for Cosmic
with nothing else on it and if they would tell me step for step how to set
it up, where to put it no wise cracks huh...lol), which command lines to
run. It's going to take me months to grasp flakes due to timeload. I
switched to Cosmic as my main driver directly from Pop. But would really
love to use it with NixOS. Need some help. Thanks to everyone so much.

**Comment by authenticastrology** [comment]
I got it all figured out! Yea!!!!!!!.

Question though....I know it's in Alpha. But is anyone else having issues with lag when things open?

I have used the Pop OS version too and it seems a bit snappier.. Not sure if it is a NixOS issue or not.

**Comment by colemickens** [comment]
If my dGPU is not active, launching windows takes longer than it feels it should. It fires up the dGPU for animations, then animates opening the new window. For me, it's quite noticeably slow, but I'm also behind on updates, so hoping it improves...

**Comment by tilupe** [comment]
Hey I just started with NixOS and wanted to install cosmic too. When running a nixos-rebuild, it always gets stuck at this phase
![image](https://github.com/user-attachments/assets/7dab5c18-6425-42fe-baa9-73745ee63e61)

It just compiles there for a long time and at some point it the terminal crashes or after a while I get logged out and everything is cleared. Anybody an idea what I can do to change that?

**Comment by hanckmann** [comment]
@tilupe I guess you are being bitten by this issue as well: #495

**Comment by artsiomkaltovich** [comment]
> one thing missing in the readme (I think?) is that you should do ` sudo nixos-rebuild test --flake .` and not just `nixos rebuild test` before adding the `cosmic enable` in your configuration.nix. Then rebuid with `sudo nixos-rebuild switch --flake .`

Unfortunattly it produces the following error for me:

       … while calling the 'seq' builtin
         at /nix/store/5wvrlpfrwlmg24ywkybvidyzcvki6bwx-source/lib/modules.nix:334:18:
          333|         options = checked options;
          334|         config = checked (removeAttrs config [ "_module" ]);
             |                  ^
          335|         _module = checked (config._module);

       … while evaluating a branch condition
         at /nix/store/5wvrlpfrwlmg24ywkybvidyzcvki6bwx-source/lib/modules.nix:273:9:
          272|       checkUnmatched =
          273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |         ^
          274|           let

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: path '/nix/store/mk0n6m7qvqd9n4smvgds403whdz9bazf-source/configuration.nix' does not exist

Have anybody experienced something like that?

**Comment by artsiomkaltovich** [comment]
> > one thing missing in the readme (I think?) is that you should do ` sudo nixos-rebuild test --flake .` and not just `nixos rebuild test` before adding the `cosmic enable` in your configuration.nix. Then rebuid with `sudo nixos-rebuild switch --flake .`
> 
> Unfortunattly it produces the following error for me:
> 
> ```
>    … while calling the 'seq' builtin
>      at /nix/store/5wvrlpfrwlmg24ywkybvidyzcvki6bwx-source/lib/modules.nix:334:18:
>       333|         options = checked options;
>       334|         config = checked (removeAttrs config [ "_module" ]);
>          |                  ^
>       335|         _module = checked (config._module);
> 
>    … while evaluating a branch condition
>      at /nix/store/5wvrlpfrwlmg24ywkybvidyzcvki6bwx-source/lib/modules.nix:273:9:
>       272|       checkUnmatched =
>       273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
>          |         ^
>       274|           let
> 
>    (stack trace truncated; use '--show-trace' to show the full, detailed trace)
> 
>    error: path '/nix/store/mk0n6m7qvqd9n4smvgds403whdz9bazf-source/configuration.nix' does not exist
> ```
> 
> Have anybody experienced something like that?

Ok, it seems the problem was that flake was initialized in my home directory and not in ``/etc/nixos``, but everything is compiled locally for me, still some progress I guess :)

**Comment by Elias-Graf** [comment]
@artsiomkaltovich no, the issue isn't where the flake flake was initialized, but that you forgot to move a file, in this case "configuration.nix" into the repository. **OR**, something I've run into a couple of times myself, flakes only consider files, that have been added to the git index. That's why my rebuild command is this:

```sh
nix fmt && git add . && sudo nixos-rebuild switch --flake .#<NAME_OF_HOST>
```

and I just reverse search my history for `rebuild`.

**Comment by RekitRalph** [comment]
Could someone take a look at my [flake.nix](https://github.com/RekitRalph/nixosConfig/blob/56045f59c6be28f4622dd1a3cf69a52dfe7c95a3/flake.nix) and help me understand why cachix is not working for me? Any help would be appreciated. I did run a rebuild before enabling the cosimc modual but it was still building cosmic when I ran the build.

**Comment by EyesWithoutFace** [comment]
Hi I am new to flakes and NixOS my self but ive had system76 before and I really like their efforts and would like to use cosmic. I modified my current flake to adjust for cosmic since I use the cachyos kernel from chaotic nyx pkgs.
I cant seem to be able do very many things with cosmic this includes:

- [ ] Use dash/dock
- [ ] use half my apps

[cosmic.log](https://github.com/user-attachments/files/19259288/cosmic.log)

- [ ] download flatpaks due to gnupg error
- [ ] apply most settings


```
{
  inputs = {
    nixpkgs.follows = "nixos-cosmic/nixpkgs"; # NOTE: change "nixpkgs" to "nixpkgs-stable" to use stable NixOS release
    chaotic.url = "github:chaotic-cx/nyx/nyxpkgs-unstable";
    nixos-cosmic.url = "github:lilyinstarlight/nixos-cosmic";
  };

  outputs = { self, nixpkgs, nixos-cosmic, chaotic }: {
    nixosConfigurations = {
      # NOTE: change "host" to your system's hostname
      NixEyes = nixpkgs.lib.nixosSystem {
        modules = [
          {
            nix.settings = {
              substituters = [ "https://cosmic.cachix.org/" ];
              trusted-public-keys = [ "cosmic.cachix.org-1:Dya9IyXD4xdBehWjrkPv6rtxpmMdRel02smYzA85dPE=" ];
            };
          }
          nixos-cosmic.nixosModules.default
          chaotic.nixosModules.default
          ./configuration.nix
        ];
      };
    };
  };
}
```

and my configuration.nix where cosmic is 


  ```
services.displayManager.cosmic-greeter.enable = true;
  services.xserver.desktopManager.gnome.enable = true;
  services.xserver.displayManager.gdm.enable = false;
  services.desktopManager.cosmic.enable = true;  #services.displayManager.cosmic-

```

I only have gnome still as a fallback 

I also attached my ` journalctl | grep cosmic > cosmic.log`

**Comment by KiaraGrouwstra** [comment]
i use the non-flake [npins](https://github.com/andir/npins) for dependency management.
while without a dependency manager you may need a function like [`fetchFromGithub`](https://nixos.org/manual/nixpkgs/stable/#fetchfromgithub) ([example](https://github.com/NixOS/nixpkgs/blob/f1a85c375555fded2c3fbb06cb7a74c64ae36e75/pkgs/by-name/co/cosmic-session/package.nix#L18-L23)) over my `"${sources.nixos-cosmic}/nixos"` reference, otherwise the flake snippet in the readme would end up something [like this](https://codeberg.org/kiara/dots/commit/10901d723815889bbde7436828419249b8edcc6d#diff-e8ee46cf7f176cbd2ae78efec7a5a8a8e28bab78).

in other words, you would be adding to your `imports`:

```nix
    (let nixos-cosmic = pkgs.fetchFromGitHub {
    owner = "lilyinstarlight";
    repo = "nixos-cosmic";
    rev = "PUT-LATEST-COMMIT-HERE";
    hash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
  }; in import "${nixos-cosmic}/nixos" { cosmicOverlay = final: prev: import "${nixos-cosmic}/pkgs" { inherit final prev; rustPlatform = prev.rustPlatform; }; })
```
